cpp/src/arrow/flight/server.cc

```diff
@@ -899,6 +921,13 @@ Status FlightServerBase::Serve() {
     ARROW_ASSIGN_OR_RAISE(old_handler, SetSignalHandler(signum, new_handler));
     impl_->old_signal_handlers_.push_back(std::move(old_handler));
   }
+#ifndef _WIN32
+  if (sem_init(&impl_->waiting_sem_, /*pshared=*/0, /*value=*/0) != 0) {
```

Comment:
Unfortunately I don't think macOS supports anonymous semaphores.
An alternative would be to use a non-blocking anonymous pipe. `write` is async-signal-safe.